File: Data-Structure/maze/maze.cpp
#include <iostream>
#include <cstring>
#include <stack>
#include <cstdio>

using namespace std;
int w, h;

const int maxn = 1000;

int maze[maxn][maxn];

struct node {
    int x,y;//1,2,3,4 下 左 上 右
};
node op, ed;

stack <node> sk;
stack <node> tmp;

bool check(const node& next) {
    if(next.x < w && next.x >= 0 &&
       next.y < h && next.y >= 0 &&
       maze[next.y][next.x] == 0) {
        return true;
    }
    return false;
}

bool path() {
    node cur; cur.x = op.x; cur.y = op.y;
    node next;
    //将起点入栈
    sk.push(cur);
    while(!sk.empty()) {
        cur = sk.top();
        //判断是否到了终点
        if(cur.x == ed.x && cur.y == ed.y)return true;
        //将该点标记为已经访问
        maze[cur.y][cur.x] = -1;
        //向下运动
        next = cur;
        next.y++;
        if(check(next)) {
            cur = next;
            sk.push(cur);
            continue;
        }
        //向左运动
        next = cur;
        next.x--;
        if(check(next)) {
            cur = next;
            sk.push(cur);
            continue;
        }
        //向上运动
        next = cur;
        next.y--;
        if(check(next)) {
            cur = next;
            sk.push(cur);
            continue;
        }
        //向右运动
        next = cur;
        next.x++;
        if(check(next)) {
            cur = next;
            sk.push(cur);
            continue;
        }
//        cur = sk.top();
        sk.pop();
//      maze[cur.x][cur.y] = 0;
    }
    return false;
}

void print() {
    for(int i = 0; i < h ; i++) {
        for(int j = 0 ; j < w ; j++) {
            cout << maze[i][j] << " ";
        }
        cout << endl;
    }
}

int main() {
    freopen("in.txt","r",stdin);
    cin >> w >> h;

    memset(maze,0,sizeof(maze));

    for(int i = 0 ; i < h ; i++) {
        for(int j = 0 ; j < w ; j++) {
            int tt;
            cin >> tt;
            if(tt == 3) {
                op.y = i;
                op.x = j;
                maze[i][j] = 0;
//                cout << op.x << " " << op.y << endl;
                continue;
            }
            if(tt == 4) {
                ed.y = i;
                ed.x = j;
                maze[i][j] = 0;
//                cout << ed.x << " " << ed.y << endl;
                continue;
            }
            maze[i][j] = tt;
        }
    }
    if(path()) {
        while(!sk.empty()) {
            node t = sk.top();
            tmp.push(t);
            sk.pop();
//            cout << t.x << " " << t.y << endl;
        }
        while(!tmp.empty()) {
            node t = tmp.top();
            tmp.pop();
            cout << t.x << " " << t.y << endl;
        }
    }
    else {
        cout << "++_++" << endl;
    }
    return 0;
}

File: Data-Structure/maze/maze.py
# This program is used to find a way out in a maze

# define node
class node(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y

sk = []
tmp = []

# get the map from input file
file = open("in.txt","r")
line1 = file.readline()
size = line1.split()
w = int(size[0])
h = int(size[1])
# print("%r\nw = %d\nh = %d\n" %(size, w, h))

# the next line shows how to init a two-demensional list
maze = [[0 for i in range(w)] for i in range(h)]
lines = file.readlines()
i = 0
for line in lines:
    j = 0
    for x in line.split():
        maze[i][j] = int(x)
        if int(x) == 3:
            op = node(i, j)
            maze[i][j] = 0
        if int(x) == 4:
            ed = node(i, j)
            maze[i][j] = 0
        j += 1
    # maze[i].append([int(x) for x in line.split()])
    i += 1

# check the next step is available or not
def check(nex):
    if(nex.x < w and nex.x >= 0 and
       nex.y < h and nex.y >= 0 and
       maze[nex.y][nex.x] == 0):
        return True
    return False

# check is there a way out or not
def path():
    cur = op
    sk.append(cur)
    while(len(sk) > 0):
        cur = sk.pop()
        sk.append(cur)
        if cur.x == ed.x and cur.y == ed.y:
            return True
        maze[cur.y][cur.x] = 6
        # go down
        nex = cur
        nex.y += 1
        if check(nex):
            cur = nex
            sk.append(cur)
            continue
        # go left
        nex = cur
        nex.x -= 1
        if check(nex):
            cur = nex
            sk.append(cur)
            continue
        # go up
        nex = cur
        nex.y -= 1
        if check(nex):
            sk.append(cur)
            continue
        # go right
        nex = cur
        nex.x += 1
        if ckeck(nex):
            sk.append(cur)
            continue
        sk.pop()
    return False

# to print the map
def pint_map():
    print("Or you can watch this map:")
    for m in maze:
        print(m)


# main
def main():
     